### ISSUE TITLE
[Exit Ticket] Week 3 — Tuesday (Josiah)

### ISSUE THREAD

**Issue by jduvalian** [joshuahopperdev/COMP-3200-TEAM-ROGUE-ONE#37]
**Week:** 3
**Class:** Tuesday

### 1. In your own words: what is the difference between `delta` and `weight_delta`?
`delta` is just the raw error `pred - goal`, and `weight_delta` is scaled by the input (`delta * input`) to have a kind of "weighted" delta tuned to the sensitivity of the input.

### 2. Code question. If `pred = 0.9`, `goal = 1.0`, and `input = 2.0`, what is `weight_delta`?
weight_delta = delta * input = (pred - goal) * input = (0.9 - 1.0) * 2.0 = -0.1 * 2.0 = -0.2

### 3. One thing from today that still feels uncertain.
I'm having trouble connecting derivatives with weight_delta. You said weight_delta is proportional to the derivative of the error with respect to weight, but what exactly *is* the derivative of the error with respect to the weight? The weight isn't directly used when calculating the error, so how could you take a derivative with respect to that? I also realize that weight_delta can be thought of the change that needs to happen to the weight, is that somewhere close to correct? (ha sorry that was more than one thing)

---

*There are no wrong answers here and I read them all. "This still feels fuzzy"
is the single most useful sentence you can send me — it's what next class gets
built around.*

**Comment by jduvalian** [comment]
@HogRed all our exit tickets are done

**Comment by HogRed** [comment]
@jduvalian 

Great question. It is totally normal to be confused here, because the weight does look like it disappears before we calculate error. But it is still there indirectly. Remember the chain of events:

prediction = input × weight
error = 1/2 × (prediction - goal)^2 <-- this is the classic formulation of the error. The 1/2 is added to cancel out the 2 from the derivative.

Now substitute the first equation into the second:

error = 1/2 × (input × weight - goal)^2

Now you can actually see the weight sitting inside the error equation. So when we ask for "the derivative of error with respect to weight," we are asking a very practical question: If I nudge this weight a tiny bit, how much will the error change, and in which direction?

That is what a derivative means here. We are not taking the derivative of the weight. We are looking at error as a function of the weight. The weight is one of the knobs that created the prediction, and the prediction created the error. Therefore changing the weight can change the error, even though there was an intermediate step between them. And this is where weight_delta comes from. With our 1/2 × squared error definition, the derivative works out to:

dError/dWeight = (prediction - goal) × input

or:

weight_delta = delta × input

So weight_delta tells us the slope of the error with respect to this weight. It tells us both how sensitive the error is to this weight and which direction is uphill.

Your last thought is very close, but I would make one small correction: weight_delta is not quite "the change that needs to happen to the weight." It tells us the direction and strength of the gradient. Gradient descent then says: new_weight = weight - learning_rate × weight_delta

So think of weight_delta as saying, "Here is the slope of the hill at this weight." The minus sign says, "Walk downhill," and the learning rate says, "Here is how large a step we're willing to take." That distinction will matter later. The gradient tells us what the landscape looks like. Gradient descent decides how we move through it.

**In class on Thursday, we will walk through the formal derivation.** 🤓
